qiskit/circuit/random/utils.py

```diff
@@ -77,81 +74,114 @@ def random_circuit(
     Raises:
         CircuitError: when invalid options given
     """
+    if num_qubits == 0:
+        return QuantumCircuit()
     if max_operands < 1 or max_operands > 3:
         raise CircuitError("max_operands must be between 1 and 3")
-
-    one_q_ops = [
-        IGate,
-        U1Gate,
-        U2Gate,
-        U3Gate,
-        XGate,
-        YGate,
-        ZGate,
-        HGate,
-        SGate,
-        SdgGate,
-        TGate,
-        TdgGate,
-        RXGate,
-        RYGate,
-        RZGate,
+    max_operands = max_operands if num_qubits > max_operands else num_qubits
+
+    gates_1q = [
+        # (Gate class, number of qubits, number of parameters)
+        (IGate, 1, 0),
+        (XGate, 1, 0),
+        (YGate, 1, 0),
+        (ZGate, 1, 0),
+        (HGate, 1, 0),
+        (SGate, 1, 0),
+        (SdgGate, 1, 0),
+        (TGate, 1, 0),
+        (TdgGate, 1, 0),
+        (RXGate, 1, 1),
+        (RYGate, 1, 1),
+        (RZGate, 1, 1),
+        (UGate, 1, 3),
+    ]
+    if reset:
+        gates_1q.append((Reset, 1, 0))
+    gates_2q = [
+        (CXGate, 2, 0),
+        (CYGate, 2, 0),
+        (CZGate, 2, 0),
+        (CHGate, 2, 0),
+        (CRZGate, 2, 1),
+        (SwapGate, 2, 0),
+        (RZZGate, 2, 1),
+        (CUGate, 2, 4),
     ]
-    one_param = [U1Gate, RXGate, RYGate, RZGate, RZZGate, CU1Gate, CRZGate]
-    two_param = [U2Gate]
-    three_param = [U3Gate, CU3Gate]
-    two_q_ops = [CXGate, CYGate, CZGate, CHGate, CRZGate, CU1Gate, CU3Gate, SwapGate, RZZGate]
-    three_q_ops = [CCXGate, CSwapGate]
+    gates_3q = [(CCXGate, 3, 0), (CSwapGate, 3, 0)]
+
+    gates = gates_1q.copy()
+    if max_operands >= 2:
+        gates.extend(gates_2q)
+    if max_operands >= 3:
+        gates.extend(gates_3q)
+    gates = np.array(
+        gates, dtype=[("class", object), ("num_qubits", np.int64), ("num_params", np.int64)]
```

Comment:
I think this is the first time I've ever seen field dtypes used in practice before. I had completely forgotten you could even do this in numpy. This is a cool application for it.

Reply:
Yeah, I only very rarely use them.  For posterity: the reasons for the choice here is that I wanted the most efficient access from `rng.choice` (which also maintains dtype in its output), and for the subsequent cumulative sums over `num_qubits` and `num_params` to have defined strides in their access patterns, so the Numpy vectorisation after the `rng` choice would all work as expected.